device_fetcher.py

```
@@ -0,0 +1,88 @@
+
+                # Make gRPC call
+                response = self.device_service_stub.List(request, metadata=self.auth_token)
+
+                # Process each device in the response
+                for device in response.result:
+                    logger.info(f"Processing device: {device.name}")
+                    devices_dict[device.name] = {
+                        "euid": device.dev_eui,
+                        "description": device.description,
+                        "device_profile_id": device.device_profile_id,
+                        "device_profile_name": device.device_profile_name,
+                        "created_at": device.created_at.ToDatetime().isoformat()
+                        if device.created_at
+                        else None,
+                        "updated_at": device.updated_at.ToDatetime().isoformat()
+                        if device.updated_at
+                        else None,
+                        "last_seen_at": device.last_seen_at.ToDatetime().isoformat()
+                        if device.last_seen_at
+                        else None,
+                    }
+
+                # Break if fewer devices than limit are returned
+                if len(response.result) < limit:
+                    break
+
+                # Increment offset for the next page
+                offset += limit
+
+                # Check if the pagination has reached the maximum limit
+                if offset >= config.MAX_DEVICES:
+                    logger.warning("Reached maximum pagination limit.")
+                    break
+
+            # Validate if devices were found
+            if not devices_dict:
+                logger.warning("No devices found for the given Application ID.")
+
+            logger.info(f"Fetched {len(devices_dict)} devices from ChirpStack.")
+            return devices_dict
+
+        except Exception as e:
+            logger.error(f"Error fetching devices: {str(e)}")
+            return {}
```

In case of an error, instead of returning an empty dictionary it is always better to either raise an exception or return an error message that indicates what went wrong. Sample code snippet:
return {"error": "Error fetching devices", "message": str(e)}